heroku/cli
# refactor(pg-v5): Move commands 'pg:pull' and 'pg:push' to CLI

## Description

Moves the `pg:pull` and `pg:push` commands to CLI, converting to OClif Core and TS.

These commands were found originally in a single file, and same for their unit tests. They've been splat into two different command files and two unit test files, with shared behavior located in a new lib file imported by both commands.

The PR also includes a fix for the usage of `async` functions on Mocha `describe` and `context` blocks that was found to cause issues polluting the global test suite: [Slack thread for reference](https://salesforce-internal.slack.com/archives/C048U82QVKQ/p1714147983332269) (Heroku internal).

## Test instructions

Before testing ensure your local PostgreSQL server installation matches the major version used currently by Heroku for their PostgreSQL databases (PostgreSQL 15.x). Alternatively, you may try creating `heroku-postgresql` databases on the same version than your local PostgreSQL installation (passing `--version=<your-local-PG-version>` when creating the database add-ons) for testing, but It might not work.

Pull this branch, install dependencies with `yarn` and build with `yarn build`. Then you can run the following commands:

### Command `pg:push`:

```
./bin/run pg:push --help                      # always test that help looks ok.
./bin/run pg:push SOURCE ADDON_NAME -a mytest-app               # pushes local database SOURCE to Heroku PG add-on database if empty.
./bin/run pg:push SOURCE ADDON_NAME -a mytest-app --exclude-table-data="TABLE1;TABLE2"       # pushes local database SOURCE to Heroku PG add-on database if empty, excluding data for tables TABLE1 and TABLE2.
```

### Command `pg:pull`:

```
./bin/run pg:pull --help                      # always test that help looks ok.
./bin/run pg:pull ADDON_NAME TARGET -a mytest-app               # pulls Heroku PG add-on database to newly created local database TARGET.
./bin/run pg:push ADDON_NAME TARGET2 -a mytest-app --exclude-table-data="TABLE1;TABLE2"       # pulls Heroku PG add-on database to newly created local database TARGET2 excluding data for tables TABLE1 and TABLE2.
```

Compare outputs against the current CLI.

## SOC2 Compliance
[GUS Work Item](https://gus.lightning.force.com/a07EE00001bDDAOYA4) (Salesforce Internal)
